checker/src/org/checkerframework/checker/minlen/MinLenTransfer.java

```diff
@@ -0,0 +1,584 @@
+                ArrayCreationNode aNode = (ArrayCreationNode) arg;
+                List<Node> args = aNode.getInitializers();
+                // if there is only one argument arg; and if arg is an array of T (T[] arg); and if T is not a primitive
+                // then array.asList(arg).size() == arg.length
+                // otherwise it is treated as varargs and array.asList(arg).size() == the number of arguments
+                if (args.size() == 1
+                        && args.get(0).getType().getKind().equals(TypeKind.ARRAY)
+                        && !((ArrayType) args.get(0).getType())
+                                .getComponentType()
+                                .getKind()
+                                .isPrimitive()) {
+                    if (args.get(0).getTree() == null) {
+                        return result;
+                    }
+                    AnnotatedTypeMirror ATM = atypeFactory.getAnnotatedType(args.get(0).getTree());
+                    AnnotationMirror anno = ATM.getAnnotation(MinLen.class);
+                    value = MinLenAnnotatedTypeFactory.getMinLenValue(anno);
+                } else {
+                    value = args.size();
+                }
+            }
+            AnnotationMirror AM = atypeFactory.createMinLen(value);
+            result.setResultValue(analysis.createSingleAnnotationValue(AM, node.getType()));
+        }
+
+        return result;
+    }
+
+    @Override
+    public TransferResult<CFValue, MinLenStore> visitArrayAccess(
+            ArrayAccessNode node, TransferInput<CFValue, MinLenStore> in) {
+        TransferResult<CFValue, MinLenStore> result = super.visitArrayAccess(node, in);
+
+        if (valueAnnotatedTypeFactory
+                .getAnnotatedType(node.getArray().getTree())
+                .hasAnnotation(ArrayLen.class)) {
+            // In this case, refine the MinLen to match the ArrayLen.
+            AnnotationMirror arrayLenAnm =
+                    valueAnnotatedTypeFactory
+                            .getAnnotatedType(node.getArray().getTree())
+                            .getAnnotation(ArrayLen.class);
+            MinLenStore store = in.getRegularStore();
+            int minlen = Collections.min(ValueAnnotatedTypeFactory.getArrayLength(arrayLenAnm));
+            Receiver rec =
+                    FlowExpressions.internalReprOf(analysis.getTypeFactory(), node.getArray());
+            store.insertValue(rec, atypeFactory.createMinLen(minlen));
+        }
+
+        return result;
+    }
+
+    /**
+     * This struct contains all of the information that the refinement functions need. It's called
+     * by each node function (i.e. greater than node, less than node, etc.) and then the results are
+     * passed to the refinement function in whatever order is appropriate for that node. Its
+     * constructor contains all of its logic. I originally wrote this for LowerBoundTransfer but I'm
```

Comment:
I spent an uncomfortable amount of time this afternoon trying to make this happen. I've cleaned up the code a bit and made it somewhat better, but it's still not really there---there's an IndexAbstractTransfer, but only the LBC is able to use it because the others need to be using their own stores and I wasn't able to figure out a good way around that. Similar problem with trying to abstract out just RefinementInfo.